src/main/java/org/passay/dictionary/MemoryMappedFileWordList.java

```diff
@@ -0,0 +1,176 @@
+   * Creates a new word list from the supplied file. The input file is read on initialization and is maintained by this
+   * class.
+   *
+   * <p><strong>NOTE</strong> Attempts to close the source file will cause {@link IOException} when {@link #get(int)} is
+   * called subsequently.</p>
+   *
+   * @param  raf  File containing words, one per line.
+   * @param  caseSensitive  Set to true to create case-sensitive word list, false otherwise.
+   *
+   * @throws  IOException  if an error occurs reading the supplied file
+   */
+  public MemoryMappedFileWordList(final RandomAccessFile raf, final boolean caseSensitive)
+    throws IOException
+  {
+    this(raf, caseSensitive, DEFAULT_CACHE_SIZE);
+  }
+
+
+  /**
+   * Creates a new word list from the supplied file. The input file is read on initialization and is maintained by this
+   * class. cachePercent is a percentage of the file size in bytes.
+   *
+   * <p><strong>NOTE</strong> Attempts to close the source file will cause {@link IOException} when {@link #get(int)} is
+   * called subsequently.</p>
+   *
+   * @param  raf  File containing words, one per line.
+   * @param  caseSensitive  Set to true to create case-sensitive word list, false otherwise.
+   * @param  cachePercent  Percent (0-100) of file to cache in memory for improved read performance.
+   *
+   * @throws  IllegalArgumentException  if cache percent is out of range.
+   * @throws  IOException  if an error occurs reading the supplied file
+   */
+  public MemoryMappedFileWordList(final RandomAccessFile raf, final boolean caseSensitive, final int cachePercent)
+    throws IOException
+  {
+    super(raf, caseSensitive, cachePercent);
+    final FileChannel channel = file.getChannel();
+    buffer = channel.map(FileChannel.MapMode.READ_ONLY, 0, channel.size());
+
+    final int fileBytes = buffer.capacity();
+    final int cacheSize = (fileBytes / HUNDRED_PERCENT) * cachePercent;
+    final int cacheOffset = cacheSize == 0 ? fileBytes : cacheSize > fileBytes ? 1 : fileBytes / cacheSize;
+
+    long pos = 0;
+    String a;
+    String b = null;
+    while (buffer.hasRemaining()) {
+      a = readLine(buffer);
+      if (b != null && comparator.compare(a, b) < 0) {
+        throw new IllegalArgumentException("File is not sorted correctly for this comparator");
+      }
+      b = a;
+
+      if (cacheSize > 0 && size % cacheOffset == 0) {
+        cache.put(size, pos);
+      }
+      pos = buffer.position();
+      size++;
+    }
+  }
+
+
+  /**
+   * Reads the file line by line and returns the word at the supplied index. Returns null if the index cannot be read.
+   * This method leverages the cache to seek to the closest position of the supplied index.
+   *
+   * @param  index  to read word at
+   *
+   * @return  word at the supplied index
+   *
+   * @throws  IllegalStateException  if an error occurs reading the supplied file
+   */
+  @Override
+  protected String readFile(final int index)
+  {
+    synchronized (buffer) {
```

Comment:
This method sets the position of the buffer, without the synchronization this method will not be thread safe.